src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/NodeFactory.cs

```diff
@@ -341,6 +341,11 @@ private void CreateNodeCaches()
 
         public IEETypeNode NecessaryTypeSymbol(TypeDesc type)
         {
+            if (!_compilationModuleGroup.IsSingleFileCompilation && EETypeNode.IsTypeNodeShareable(type))
```

Comment:
Shouldn't this be just conditioned on `ShouldProduceFullType`? It feels like this is inlining logic that should be decided in the compilation group.